# Проект по SQL
Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Наша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Наша первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Импорт-библиотек-и-данных" data-toc-modified-id="Импорт-библиотек-и-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Импорт библиотек и данных</a></span></li><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Описание данных</a></span></li><li><span><a href="#Анализ-данных-и-расчёты" data-toc-modified-id="Анализ-данных-и-расчёты-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Анализ данных и расчёты</a></span><ul class="toc-item"><li><span><a href="#Посчитайте,-сколько-книг-вышло-после-1-января-2000-года" data-toc-modified-id="Посчитайте,-сколько-книг-вышло-после-1-января-2000-года-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Посчитайте, сколько книг вышло после 1 января 2000 года</a></span></li><li><span><a href="#Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку" data-toc-modified-id="Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Для каждой книги посчитайте количество обзоров и среднюю оценку</a></span></li><li><span><a href="#Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц" data-toc-modified-id="Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Определите издательство, которое выпустило наибольшее число книг толще 50 страниц</a></span></li><li><span><a href="#Определите-автора-с-самой-высокой-средней-оценкой-книг" data-toc-modified-id="Определите-автора-с-самой-высокой-средней-оценкой-книг-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Определите автора с самой высокой средней оценкой книг</a></span></li><li><span><a href="#Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок" data-toc-modified-id="Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></div>

## Импорт библиотек и данных

In [122]:
import pandas as pd
from sqlalchemy import create_engine

In [123]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [124]:
def read_sql(SQL):
    '''Функция передает запрос SQL методу read_sql() pandas.
    '''
    return pd.io.sql.read_sql(SQL, con = engine)

## Описание данных

In [125]:
SQL = 'SELECT * FROM books LIMIT 5'
read_sql(SQL)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

Таблица books

Содержит данные о книгах:
-	`book_id` — идентификатор книги;
-	`author_id` — идентификатор автора;
-	`title` — название книги;
-	`num_pages` — количество страниц;
-	`publication_date` — дата публикации книги;
-	`publisher_id` — идентификатор издателя.

In [126]:
SQL = 'SELECT * FROM authors LIMIT 5'
read_sql(SQL)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

Таблица authors

Содержит данные об авторах:
-	`author_id` — идентификатор автора;
-	`author` — имя автора.

In [127]:
SQL = 'SELECT * FROM publishers LIMIT 5'
read_sql(SQL)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

Таблица publishers

Содержит данные об издательствах:
-	`publisher_id` — идентификатор издательства;
-	`publisher` — название издательства;

In [128]:
SQL = 'SELECT * FROM ratings LIMIT 5'
read_sql(SQL)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

Таблица ratings

Содержит данные о пользовательских оценках книг:
-	`rating_id` — идентификатор оценки;
-	`book_id` — идентификатор книги;
-	`username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

In [129]:
SQL = 'SELECT * FROM reviews LIMIT 5'
read_sql(SQL)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

Таблица reviews

Содержит данные о пользовательских обзорах:
-	`review_id` — идентификатор обзора;
-	`book_id` — идентификатор книги;
-	`username` — имя автора обзора;
-	`text` — текст обзора.

**Вывод:** все таблицы загружены, отображаются корректно.

## Анализ данных и расчёты

Нам необходимо выполнить следующие задания:
-	Посчитайте, сколько книг вышло после 1 января 2000 года;
-	Для каждой книги посчитайте количество обзоров и среднюю оценку;
-	Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
-	Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
-	Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

### Посчитайте, сколько книг вышло после 1 января 2000 года

In [130]:
read_sql('''
SELECT
    COUNT(DISTINCT book_id) AS books_cnt 
FROM
    books 
WHERE
    publication_date > '2000-01-01'
''')

books_cnt
0        819

После 1 января 2000 года вышло 819 книг.

### Для каждой книги посчитайте количество обзоров и среднюю оценку

Нам необходимо соединить таблицы `reviews` и `ratings`, произвести расчеты, для удобства визуализации присоединим также таблицу `books` с названиями произведений:

In [131]:
read_sql('''
SELECT
    reviews.book_id,
    books.title AS title,
    COUNT(DISTINCT(review_id)) AS review_count,
    AVG(rating) AS avg_rating
FROM
    ratings
    LEFT JOIN reviews ON reviews.book_id = ratings.book_id
    LEFT JOIN books ON books.book_id = ratings.book_id
GROUP BY
    reviews.book_id,
    books.title
ORDER BY
    review_count DESC,
    avg_rating DESC
    LIMIT 10
''')

book_id                                              title  review_count  \
0      948                            Twilight (Twilight  #1)             7   
1      302  Harry Potter and the Prisoner of Azkaban (Harr...             6   
2      299  Harry Potter and the Chamber of Secrets (Harry...             6   
3      656                                     The Book Thief             6   
4      734                                   The Glass Castle             6   
5      497                          Outlander (Outlander  #1)             6   
6      750                The Hobbit  or There and Back Again             6   
7      695  The Curious Incident of the Dog in the Night-Time             6   
8      779  The Lightning Thief (Percy Jackson and the Oly...             6   
9      963                                Water for Elephants             6   

   avg_rating  
0    3.662500  
1    4.414634  
2    4.287500  
3    4.264151  
4    4.206897  
5    4.125000  
6    4.125000  
7    4.081081  
8    4.080645  
9    3.977273

По количеству обзоров лидер "Сумерки", однако средняя оценка этой книги - довольно "средняя" - 3,7.

Мы отфильтровали список по двум параметрам (обзоры и средняя оценка), получили список топ-10 книг. На 2 и 3 месте - книги о Гарри Поттере.

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц

Соединим таблицы `books` и `publishers`, посчитаем количество выпущенных книг и выведем топ-3 издательства:

In [132]:
read_sql('''
SELECT
    publishers.publisher, 
    COUNT(book_id) AS book_count
FROM 
    books 
    INNER JOIN publishers ON publishers.publisher_id=books.publisher_id
WHERE 
    num_pages > 50 
GROUP BY 
    publishers.publisher_id 
ORDER BY 
    book_count DESC 
    LIMIT 3
''')

publisher  book_count
0             Penguin Books          42
1                   Vintage          31
2  Grand Central Publishing          25

- на 1 месте `Penguin Books` - 42 книги
- на 2 месте `Vintage` - 31 книга
- на 3 месте `Grand Central Publishing` - 25 книг

### Определите автора с самой высокой средней оценкой книг

Посмотрим топ-5 авторов с самой высокой средней оценкой:

In [133]:
read_sql('''
WITH top_author AS
    (SELECT
        book_id,
        COUNT(rating_id) AS count,
        AVG(ratings.rating) AS avg_rating
    FROM
        ratings
    GROUP BY
        book_id
    HAVING
        COUNT(rating_id) >= 50)
        
SELECT
    authors.author,
    AVG(top_author.avg_rating) AS rating
FROM
    books
    INNER JOIN authors ON authors.author_id = books.author_id
    INNER JOIN top_author ON top_author.book_id = books.book_id
GROUP BY
    authors.author,
    authors.author_id
ORDER BY
    rating DESC
    LIMIT 5
''')

author    rating
0         J.K. Rowling/Mary GrandPré  4.283844
1  Markus Zusak/Cao Xuân Việt Khương  4.264151
2                     J.R.R. Tolkien  4.258446
3                  Louisa May Alcott  4.192308
4                       Rick Riordan  4.080645

- На 1 месте писательница `J.K. Rowling` - автор книг о Гарри Поттере - перекликается с рейтингом книг (входят в топ-3).
- на 2 месте `Markus Zusak` - автор "Книжный вор" (топ 4 нашего списка)
- на 3 месте `J.R.R. Tolkien` - автор "Властелина колец" и "Хоббит: туда и обратно" (топ 7 из нашего списка)

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [134]:
read_sql('''
SELECT
    AVG(reviews_num) AS avg_count
FROM
    (SELECT
        COUNT(review_id) AS reviews_num
    FROM
        reviews
        INNER JOIN
        (SELECT
            username,
            COUNT (rating_id) AS ratings_num
        FROM
            ratings
        GROUP BY
            username
        HAVING COUNT (rating_id) > 50 ) AS ratings_users ON ratings_users.username = reviews.username
    GROUP BY
        reviews.username) AS reviews_num
''')

avg_count
0  24.333333

Пользователи, которые поставили больше 50 оценок, в среднем делают 24 обзора.

## Вывод

- После 1 января 2000 года вышло `819 книг`.
- По количеству обзоров лидирует произведение `"Сумерки"`, средняя оценка этой книги - 3,7.
- Наибольшее число книг толще 50 страниц выпустило издательство `Penguin Books` - 42 книги.
- Автор с самой высокой средней оценкой книг - `J.K. Rowling` - 4,28.
- Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - `24 обзора`.